### PR TITLE
Adding welcome bot, and refactoring some messy things

### PR DESCRIPTION
**Pull request by sesamechicken**
Cleans up the displayMessage function, creates a utils.js file for utitilty functions, adds the welcome bot for new users

### CODE REVIEW

**Comment by sesamechicken** [comment]
Addresses https://github.com/KamandPrompt/baat-cheet/issues/144

File: public/utils.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+/**
+ * Method to convert an array into an unordered list
+ * @param {array} arr Array of items
+ */
+const convertIntoList = (arr) => {
+  let list = ('<ul>');
+  for(let i=0; i<arr.length; i++)	list = list.concat(`<li>${arr[i]}</li>`);
+  list = list.concat('</ul>');
+  return list;
+}
```
Comment (LalitNM) [inline public/utils.js:17]: You should add all these functions in `script.js` and rename that to `utils.js`.

File: public/client.js [hunk 2]
```diff
@@ -281,16 +196,16 @@ socket.on('room created other', (data) =>{
...
-                                      ${emobox}
+                                        <i class="far fa-smile icon"></i>
+                                        ${emobox}
                                     </button>
                                     <button class="btn send text-primary" type="button">
-                                      <i class="fas fa-paper-plane icon"></i>
+                                        <i class="fas fa-paper-plane icon"></i>
                                     </button>
-                                  </div>
+                                </div>
```
Comment (LalitNM) [inline public/client.js]: Can you just remove all these format changes.

File: app.js [hunk 3]
```diff
@@ -59,6 +61,12 @@ io.on('connection', (socket) => {
 		}
 	});
 
+
+	// Welcome the user to the app
+	const welcomeUser = (socket, data) => {
+		io.to(socket.id).emit('welcome user', data)
+	}
+
```
Comment (LalitNM) [inline app.js:81]: Do not make a separate socket for welcome-bot. Just add a user named 'WelcomeBot' and use this user to welcome participants on joining. I want to keep it similar to bots on a discord server.

File: app.js [hunk 4]
```diff
@@ -147,10 +155,10 @@ io.on('connection', (socket) => {
...
 				num_users = rooms[i].num_users;
 				for( let j = 0; j < rooms[i].users.length; j++){
-				   if ( rooms[i].users[j] === socket.username) {
-				     rooms[i].users.splice(j, 1);
-				     j--;
-				   }
+					if ( rooms[i].users[j] === socket.username) {
+						rooms[i].users.splice(j, 1);
+						j--;
+					}
```
Comment (LalitNM) [inline app.js]: These are just spaces converted into tabs by your text editor. Remove these unnecessary changes.

File: public/client.js [hunk 5]
```diff
@@ -95,16 +78,16 @@ const appendContentInfo = (room_name, online, data_joined) => {
...
+                                    <textarea type="text" class="textarea form-control" placeholder="Message to ${room_name}..." data-active="lobby" rows="2"></textarea>
+                                    <div class="input-group-append">
+                                        <button class="btn smiley text-primary" type="button">
+                                            <i class="far fa-smile icon"></i>
+                                            ${emobox}
+                                        </button>
+                                        <button class="btn send text-primary" type="button">
+                                            <i class="fas fa-paper-plane icon"></i>
+                                        </button>
+                                    </div>
```
Comment (LalitNM) [inline public/client.js]: Again some format changes, revert these too.

File: public/client.js [hunk 6]
```diff
@@ -155,97 +145,22 @@ socket.on('user join', (data) => {
     }
 });
 
-//displays message to users
+// displays message to users
+// TODO: this method is GIANT and should be broken into smaller parts
+// It's also very brittle, relying on the format of the msg
```
Comment (LalitNM) [inline public/client.js]: Now you have achieved this task, so you can remove these comments.

File: public/script.js [hunk 7]
```diff
@@ -73,11 +71,11 @@ function collap(room_id) {
...
   if(height == '0') {
-   $(`#${room_id}-msg`).find('.Participants').css({"opacity":"1" , "z-index":"10"});
-  $(`#${room_id}-msg`).find('.btn').addClass('viewUsers');
+    $(`#${room_id}-msg`).find('.Participants').css({"opacity":"1" , "z-index":"10"});
+    $(`#${room_id}-msg`).find('.btn').addClass('viewUsers');
   } else {
-   $(`#${room_id}-msg`).find('.Participants').css({"opacity":"0" , "z-index":"-10"});
-  $(`#${room_id}-msg`).find('.btn').removeClass('viewUsers');
+    $(`#${room_id}-msg`).find('.Participants').css({"opacity":"0" , "z-index":"-10"});
+    $(`#${room_id}-msg`).find('.btn').removeClass('viewUsers');
```
Comment (LalitNM) [inline public/script.js]: You should change the variable `height` to `opacity`, and remove unnecessary styling changes.

File: public/script.js [hunk 8]
```diff
@@ -178,3 +181,67 @@ function notify(data, type){
...
+              msg = msg.slice(0, colon1) + "<img class=\"emoji\" src=\"images/emoji/" + emojis[position] + ".png\">" + msg.slice(colon2 + 1);
+          }
+          colon1 = msg.indexOf(":", colon2 + 1);
+      } else {
+          break;
+      }
+    }
+
+    // Format the current timestamp
+    const timestamp = dateFns.format(new Date(), 'H:mm:ss MMM DD');
```
Comment (LalitNM) [inline public/script.js:212]: Can we format our timestamp just like the gitter does, it is very clean and there no date is displayed if the message was received today, and for all past days it displays month, date and time. See screenshots

Message sent today:
image
Message sent in past days:
image

And also place the timestamp in front of username, like in the image.

**Review by LalitNM — CHANGES_REQUESTED** [review]
Everything else looks good, make required changes then I will again review your pull request.

File: app.js (hunk 3 above)
Comment (sesamechicken) [reply app.js:81]: It's not a separate socket, it's the main socket instance passed into the function.

File: app.js (hunk 4 above)
Comment (sesamechicken) [reply app.js]: I'll do what I can. Maybe there should be an `.editorconfig` file in this project?

File: public/client.js (hunk 6 above)
Comment (sesamechicken) [reply public/client.js]: 👍

File: public/script.js (hunk 7 above)
Comment (sesamechicken) [reply public/script.js]: I never touched this code, thus this may be some scope creep.

File: public/script.js (hunk 8 above)
Comment (sesamechicken) [reply public/script.js:212]: We can, but I think this is scope creep and a new issue should be created for it.

File: public/utils.js (hunk 1 above)
Comment (sesamechicken) [reply public/utils.js:17]: I'm not understanding what you're asking. If there are methods that should go in `utils.js`, please let me know. I don't think everything in `script.js` is a utility function, is it?

Comment (LalitNM) [reply public/utils.js:17]: At present, there are few lines of code in `script.js` which are not utility function(that's why I created #117), but that file is there just to add utility functions.

File: app.js (hunk 3 above)
Comment (LalitNM) [reply app.js:81]: I can see that, but I wanted to say that you should just make a permanent user of the app named `WelcomeBot`, and treat it as another person using the app.

File: app.js (hunk 4 above)
Comment (LalitNM) [reply app.js]: I am not in favor of something as strong as  `.editorconfig`, but some guidelines should be there about the codebase in `CONTRIBUTING.md`.

File: public/script.js (hunk 7 above)
Comment (LalitNM) [reply public/script.js]: I understand that. Leave it for now.

File: public/script.js (hunk 8 above)
Comment (LalitNM) [reply public/script.js:212]: I agree with you that there should be a new issue for this. I pointed out this because you added `date-fns`. Leave this for now. I will create another issue for this. But only after merging the `date-fns` feature.

File: app.js (hunk 3 above)
Comment (sesamechicken) [reply app.js:81]: So you will have many users in your app that potentially say 1 thing and never again are used? I think we should talk about a more efficient way to approach this. One bot to communicate with a user makes more sense.

File: public/client.js (hunk 5 above)
Comment (sesamechicken) [reply public/client.js]: Fixed

File: public/utils.js (hunk 1 above)
Comment (sesamechicken) [reply public/utils.js:17]: Again, we're introducing scope creep. I pulled out some methods into the `utils.js` file because they seemed re-usable. If you want all of them in a separate file, then another issue should be made for this, IMO. It has nothing to do with the welcome bot.

File: public/client.js (hunk 2 above)
Comment (sesamechicken) [reply public/client.js]: Fixed

**Comment by LalitNM** [comment]
@sesamechicken, you need to rebase your branch with `KamdndPrompt: master`.

**Comment by sesamechicken** [comment]
@LalitNM Added the Welcome Bot as a proper user to the app.

File: public/script.js [hunk 9]
```diff
@@ -165,3 +172,67 @@ function notify(data, type) {
...
+    }
+
+    // Adding Emoji
+    let p;
+    let colon1 = msg.indexOf(":");
+    while (colon1 != -1) {
+      let colon2 = msg.indexOf(":", colon1 + 1);
+      if (colon2 != -1) {
+          emoji_name = msg.slice(colon1 + 1, colon2);
+          position = emoji_names.indexOf(emoji_name)
```
Comment (LalitNM) [inline public/script.js]: This code has been changed after merging #147. You need to take care of this. With this code emojis are not working now,.

File: views/footer_imports.ejs [hunk 10]
```diff
@@ -1,6 +1,8 @@
+<script src="https://cdnjs.cloudflare.com/ajax/libs/date-fns/1.30.1/date_fns.min.js" integrity="sha512-F+u8eWHrfY8Xw9BLzZ8rG/0wIvs0y+JyRJrXjp3VjtFPylAEEGwKbua5Ip/oiVhaTDaDs4eU2Xtsxjs/9ag2bQ==" crossorigin="anonymous"></script>
+<script src='/utils.js'></script>
 <script src='https://cdnjs.cloudflare.com/ajax/libs/jquery/3.2.1/jquery.min.js'></script>
 <script src="https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/js/bootstrap.min.js" integrity="sha384-JjSmVgyd0p3pXB1rRibZUAYoIIy6OrQ6VrjIEaFf/nJGzIxFDsf4x0xIM+B07jRM" crossorigin="anonymous"></script>
 <script src="https://cdnjs.cloudflare.com/ajax/libs/socket.io/1.4.5/socket.io.js"></script>
 <script type="text/javascript" src="/emoji.json"></script>
 <script src='/script.js'></script>
-<script src="/client.js"></script>
+<script src="/client.js"></script>
```
Comment (LalitNM) [inline views/footer_imports.ejs]: There must be an extra line at the end of each file(according to `.editorconfig`.

File: public/client.js [hunk 11]
```diff
@@ -272,60 +197,63 @@ socket.on('room created other', (data) => {
...
-	var room_id = convertIntoId(name);
-	$(`#${room_id}-msg`).find('.top').find('span')[1].innerHTML = online + " user(s) online";
-	$(`#${room_id}-msg`).find('.Participants').find('span')[0].innerHTML = convertIntoList(online_users);
+
+// updates info about number of users
+socket.on('room joined', function(data) {
+    const { name, online, online_users } = data;
+    var room_id = convertIntoId(name);
+    $(`#${room_id}-msg`).find('.top').find('span')[1].innerHTML = online + " user(s) online";
+    $(`#${room_id}-msg`).find('.Participants').find('span')[0].innerHTML = convertIntoList(online_users);
```
Comment (LalitNM) [inline public/client.js:258]: Don't you think that this is unnecessary(according to `.editorconfig`.

Comment (sesamechicken) [reply public/client.js:258]: Im unsure what you're talking about. I didn't touch that portion of code.

File: views/footer_imports.ejs (hunk 10 above)
Comment (sesamechicken) [reply views/footer_imports.ejs]: Done

File: public/script.js (hunk 9 above)
Comment (sesamechicken) [reply public/script.js]: Sigh...we need unit tests in this project desperately.

Comment (LalitNM) [reply public/script.js]: Yeah, I am about to merge one pull request for this.

Comment (sesamechicken) [reply public/script.js]: This is fixed.

**Comment by sesamechicken** [comment]
@LalitNM I think that's everything. 😉

File: public/client.js [hunk 12]
```diff
@@ -17,16 +17,16 @@ const setUsername = () => {
...
-		msg: msg,
-		room: room_name
-	});
+const sendMessage = (msg)=>  {
+    msg = msg.replace(/</g, "&lt;").replace(/>/g, "&gt;");
+    room_name = $(".active").attr("id");
+    socket.emit('Message Request', {
+        msg: msg,
+        room: room_name
+    });
```
Comment (LalitNM) [inline public/client.js:26]: If you haven't changed this code, then why it is here :sweat_smile:

File: public/client.js [hunk 13]
```diff
@@ -92,16 +84,16 @@ const appendContentInfo = (room_name, online, data_joined) => {
...
                                       ${emobox}
-                                    </button>
-                                    <button class="btn send text-primary" type="button">
+                                      </button>
+                                      <button class="btn send text-primary" type="button">
                                       <i class="fas fa-paper-plane icon"></i>
-                                    </button>
-                                  </div>
+                                      </button>
+                              </div>
```
Comment (LalitNM) [inline public/client.js:96]: @sesamechicken you need to follow `.editorconfig` strictly, and if there is some ambiguity there please let me know or update that file also.

**Review by LalitNM — COMMENTED** [review]
Your pull request is good, but there is a lot unnecessary changes here. I would suggest to reset your branch to few commits and then rebase and then again make changes, this way you can get rid of all the unnecessary changes. Or remove these changes one by one.

File: public/client.js (hunk 12 above)
Comment (sesamechicken) [reply public/client.js:26]: It's there because it's been formatted. You need to decide on how strict your formatting is going to be and what you're going to use to enforce it. ESLint can do this, as can `prettier`. But, the code itself is the same.

File: public/client.js (hunk 13 above)
Comment (sesamechicken) [reply public/client.js:96]: I didn't go against the grain on the editor config. Again, there seems to be some misunderstanding surrounding linting/`eslint` and editor config. Happy to chat about it, but this is it for me on this one.

**Review by LalitNM — APPROVED** [review]

**Comment by LalitNM** [comment]
Merged @sesamechicken :rocket:
